src/SonarQube.Client/Services/SonarQubeClient.cs

```diff
@@ -0,0 +1,284 @@
+﻿/*
+ * SonarQube Client
+ * Copyright (C) 2016-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Web;
+using Google.Protobuf;
+using Newtonsoft.Json.Linq;
+using SonarQube.Client.Helpers;
+using SonarQube.Client.Messages;
+
+namespace SonarQube.Client.Services
+{
+    public class SonarQubeClient : ISonarQubeClient
+    {
+        private readonly HttpMessageHandler messageHandler;
+        private readonly TimeSpan requestTimeout;
+
+        public SonarQubeClient(HttpMessageHandler messageHandler, TimeSpan requestTimeout)
+        {
+            if (messageHandler == null)
+            {
+                throw new ArgumentNullException(nameof(messageHandler));
+            }
+
+            if (requestTimeout <= TimeSpan.Zero)
+            {
+                throw new ArgumentException("Doesn't expect a zero or negative timeout.", nameof(requestTimeout));
+            }
+
+            this.messageHandler = messageHandler;
+            this.requestTimeout = requestTimeout;
+        }
+
+        public Task<Result<ComponentResponse[]>> GetComponentsSearchProjectsAsync(ConnectionRequest connection,
+                                            ComponentRequest request, CancellationToken token)
+        {
+            const string SearchProjectsAPI = "api/components/search_projects"; // Since 6.2; internal
+
+            var query = AppendQueryString(SearchProjectsAPI, "?organization={0}&p={1}&ps={2}&asc=true",
+                        request.OrganizationKey, request.Page, request.PageSize); // TODO: should handle optional params
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this 'TODO' comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)